developer_support_script/check_build_stability.py

```diff
@@ -1,45 +1,162 @@
 
-test_metadata = "https://epics-jenkins.isis.rl.ac.uk/job/System_Tests/api/json"
+class SystemTestData:
+    """
+    Singleton class to host system test data pulled down from jenkins
+    """
+
+    def __init__(self, system_tests_url, test_metadata, test_return_quantity):
+        self.system_tests_url = system_tests_url
+        self.raw_test_metadata = test_metadata
+        self.test_return_quantity = test_return_quantity
+
+        self._json_test_metadata = self.request_page_and_metadata()
+        self._raw_builds = self.get_all_builds()
+        self._current_build, self._complete_build = self.get_current_and_complete_builds()
+        self._proccessed_builds = self.remove_current_build_if_no_results()
+        self.failing_tests = self.retrieve_failed_tests_from_builds()
+        self.top_n_failing_tests = self.get_top_n_failed_tests()
+
+    def request_page_and_metadata(self):
+        """
+        Request a jenkins webpage and return
+        the metadata json containing system test build info
+
+        :return: the metadata json
+        """
+        try:
+            page = urllib.request.urlopen(self.raw_test_metadata)
+            metadata = json.loads(page.read())
+            self._json_test_metadata = metadata
+            return metadata
+        except http.IncompleteRead as err:
+            err_page = err.partial
+            print(f"Failed to get metadata: {err_page}")
+            return None
+
+    def get_all_builds(self):
+        """
+        Return a list of all builds in the system tests job
+
+        :return: a list of all builds in the system tests job
+        """
+        builds = [build["number"] for build in self._json_test_metadata["builds"]]
+        self._raw_builds = builds
+        return builds
+
+    def get_current_and_complete_builds(self):
+        """
+        Return the current and complete builds from the metadata json
+
+        :return: the current and complete builds from the metadata json
+        """
+
+        current_build =  self._json_test_metadata["lastBuild"]["number"]
+        complete_build =  self._json_test_metadata["lastCompletedBuild"]["number"]
+
+        return current_build, complete_build
+
+    def remove_current_build_if_no_results(self):
+        """
+        Remove the current build from the list of builds if there
+        are no results for it
+
+        :return: the list of builds without the current build if there
+        """
+        if self._complete_build != self._current_build:
+            self._raw_builds.remove( self._current_build)
+        return self._raw_builds
+
+    def retrieve_test_data(self, test_num: int):
+        """
+        Retrieve test data for a given build number
+
+        :param test_num: the build number to retrieve test data for
+        :return: the test data json
+        """
+        print(f"Getting data for {test_num}")
+        try:
+            page = urllib.request.urlopen(self.system_tests_url.format(test_num))
```

Comment:
Same as I mentioned above, pylin seems to prefer use of with, though Im not sure it actually matters.